### PR TITLE
#161966910 Implement User Registration Feature

### PR DESCRIPTION
**Pull request by ThaDeveloper**
#### What does this PR do?
- Implement user registration feature
#### Description of Task to be completed?
- implement user registration endpoint
- Use regex to ensure data passed by the user meets the criteria set.
#### How should this be manually tested?
1. `git clone -b ft-user-registration-162541314 https://github.com/andela/ah-infinity-stones` and `$ cd ah-infinity-stones`
2. Create and activate virtual env
- `virtualenv  -p python3 venv`
- `$ source venv/bin/activate`
3. `$ pip3 install -r requirements.txt` to install the dependencies
4. Setup environment variables. Copy all from below and paste them into the .env file. Update the values to your choice.
```export DATABASE_NAME='ah_dev'
export DATABASE_USER='user'
export DATABASE_PASSWORD='password'
export HOST='localhost'
export PORT='5432'
export DATABASE_TEST='ah_test'
```
Activate the env variables by `$source .env`.
5. Run migrations: `$ python3 manage.py migrate`
6. Run app - `$ python3 manage.py runserver`
7. Acess `http://127.0.0.1:8000/api/users/` through [Postman](https://www.getpostman.com/), provide inputs in the following format and submit.
```
{ "user": {
"username": "username44",
"email": "username@user.com",
"password": "@Password321"	
}
}
```
8. Run tests`python3 manage.py test authors.apps.authentication.tests.test_registration`
#### What are the relevant pivotal tracker stories?
[#161966910](https://www.pivotaltracker.com/story/show/161966910)
#### Screenshots
screenshot 2018-12-13 at 10 57 28

### CODE REVIEW

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -1,10 +1,45 @@
...
+            user = User.objects.get(email=payload["email"])
+        except User.DoesNotExist:
+            raise AuthenticationFailed('No user found for token provided')
+
+        # Check whether the user is active
+        if not user.is_active:
+            raise AuthenticationFailed("Your account is disabled, please visit your account to activate")
+        return (user, token)
+
+    def generate_token(self, email):
```
Comment (codeplus254) [inline authors/apps/authentication/backends.py:34]: At what point do you use this function?

Comment (ThaDeveloper) [reply authors/apps/authentication/backends.py:34]: when registering a new user a token needs to be generated

File: authors/apps/authentication/backends.py [hunk 2]
```diff
@@ -1,10 +1,45 @@
...
-#
-# from rest_framework import authentication, exceptions
-#
-# from .models import User
+import jwt
+from django.conf import settings
+from rest_framework import authentication, exceptions
+from rest_framework.exceptions import AuthenticationFailed
+from datetime import datetime, timedelta
+from .models import User
```
Comment (r-wambui) [inline authors/apps/authentication/backends.py:7]: Add another blank line before creating a class

File: authors/apps/authentication/backends.py [hunk 3]
```diff
@@ -1,10 +1,45 @@
-# import jwt
-#
-# from django.conf import settings
-#
-# from rest_framework import authentication, exceptions
-#
-# from .models import User
+import jwt
+from django.conf import settings
+from rest_framework import authentication, exceptions
```
Comment (r-wambui) [inline authors/apps/authentication/backends.py:3]: Remove exceptions since it has not been used anywhere.

**Pull request merged**